backend/pkg/api/types/data_access.go

```diff
@@ -331,3 +352,169 @@ const CtxUserIdKey CtxKey = "user_id"
 const CtxIsMockedKey CtxKey = "is_mocked"
 const CtxMockSeedKey CtxKey = "mock_seed"
 const CtxDashboardIdKey CtxKey = "dashboard_id"
+
+// -------------------------
+// Search structs
+
+// -- General
+type SearchType int
+
+// all possible search types
+const (
+	SearchTypeInteger SearchType = iota
+	SearchTypeName
+	SearchTypeEthereumAddress
+	SearchTypeWithdrawalCredential
+	SearchTypeEnsName
+	SearchTypeGraffiti
+	SearchTypeEmail
+	SearchTypePassword
+	SearchTypeEmailUserToken
+	SearchTypeJsonContentType
+	// Validator Dashboard
+	SearchTypeValidatorDashboardPublicId
+	SearchTypeValidatorPublicKeyWithPrefix
+	SearchTypeValidatorPublicKey
+)
+
+type Searchable interface {
+	GetSearches() []SearchType
+	SetSearchValue(s string)
+	SetSearchType(st SearchType, b bool)
+	IsEnabled() bool
+	HasAnyMatches() bool
+}
+
+// not to be used directly, only for embedding
+// think of this as a base class which provides default implementations that can be overridden / shadowed / masked (see below)
+type basicSearch struct {
+	types map[SearchType]bool
+	value string
+}
+
+func (bs *basicSearch) SetSearchValue(s string) {
+	if bs == nil {
+		log.Warnf("BasicSearch is nil, can't apply search: %s", s)
+		return
+	}
+	bs.value = s
+}
+
+func (bs *basicSearch) SetSearchType(st SearchType, b bool) {
+	if bs.types == nil {
+		bs.types = make(map[SearchType]bool)
+	}
+	bs.types[st] = b
+}
+
+func (bs *basicSearch) IsEnabled() bool {
+	return bs != nil && bs.value != ""
+}
+
+func (bs *basicSearch) HasAnyMatches() bool {
+	for _, v := range bs.types {
+		if v {
+			return true
+		}
+	}
+	return false
+}
+
+func (bs *basicSearch) GetSearches() []SearchType {
+	return []SearchType{
+		SearchTypeName,
+		SearchTypeInteger,
+		SearchTypeEthereumAddress,
+		SearchTypeWithdrawalCredential,
+		SearchTypeEnsName,
+		SearchTypeGraffiti,
+		SearchTypeEmail,
+		SearchTypePassword,
+		SearchTypeEmailUserToken,
+		SearchTypeJsonContentType,
+		SearchTypeValidatorDashboardPublicId,
+		SearchTypeValidatorPublicKeyWithPrefix,
+		SearchTypeValidatorPublicKey,
+	}
+}
+
+type baseSearchResult[T any] struct {
+	Enabled bool
+	Value   T
+}
+
+type SearchNumber baseSearchResult[uint64]
+type SearchString baseSearchResult[string]
+
+func (bs *basicSearch) AsNumber(st SearchType) SearchNumber {
+	if !bs.IsEnabled() {
+		return SearchNumber{}
+	}
+
+	if !bs.types[st] {
+		log.Warn("tried accessing invalid search: ", st)
+		return SearchNumber{}
+	}
+
+	switch st {
+	case SearchTypeInteger:
+		number, err := strconv.ParseUint(bs.value, 10, 64)
+		if err != nil {
+			log.Error(err, "error converting search value, check regex parsing", 0)
```

Comment:
this should definitely return the error and not just log it

Reply:
I tried to explicitly design the entire validation process in a way that api handles errors early so that data access doesn't have to deal with catching anymore at the point of using the result, as this is kind of the point of everything and greatly improves readability. Converting the raw string result to the expected format should happen earlier though, I'll try to revert back to using struct fields instead of methods